include/deal.II/lac/petsc_ts.h

```diff
@@ -0,0 +1,706 @@
+//-----------------------------------------------------------
+//
+//    Copyright (C) 2022 by the deal.II authors
+//
+//    This file is part of the deal.II library.
+//
+//    The deal.II library is free software; you can use it, redistribute
+//    it, and/or modify it under the terms of the GNU Lesser General
+//    Public License as published by the Free Software Foundation; either
+//    version 2.1 of the License, or (at your option) any later version.
+//    The full text of the license can be found in the file LICENSE.md at
+//    the top level directory of deal.II.
+//
+//---------------------------------------------------------------
+//
+// Author: Stefano Zampini, King Abdullah University of Science and Technology.
+
+#ifndef dealii_petsc_ts_h
+#define dealii_petsc_ts_h
+
+#include <deal.II/base/config.h>
+
+#ifdef DEAL_II_WITH_PETSC
+#  include <deal.II/base/mpi.h>
+#  include <deal.II/base/parameter_handler.h>
+#  include <deal.II/base/smartpointer.h>
+
+#  include <deal.II/lac/petsc_matrix_base.h>
+#  include <deal.II/lac/petsc_precondition.h>
+#  include <deal.II/lac/petsc_vector_base.h>
+
+#  include <petscdm.h>
+#  include <petscts.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+// Shorthand notation for PETSc error codes.
+#  define AssertTS(code)                                 \
+    do                                                   \
+      {                                                  \
+        PetscErrorCode __ierr = (code);                  \
+        AssertThrow(__ierr == 0, ExcPETScError(__ierr)); \
+      }                                                  \
+    while (0)
```

Comment:
Since we're in a header file, I'm not excited about defining non-namespaced macros. 

Our usual style is to move the implementation of all member functions out-of-line and into a `.cc` file. Would this be something I could talk you into? Then the macro could also be moved to the `.cc` file.

Reply:
I will let @luca-heltai answer for the abomination then :-)

But in seriousness: As far as I can see, if all of the implementations were put into a `.cc` file, then nothing in the `.h` file actually depends on anything related to PETSc. We could put the `#include <petsc*.h>` into the `.cc` file and the `AssertTS` would fit there as well. This would surely also improve compile times. So I have a fairly strong preference for this.

Reply:
AFAIK, with the cc file you will need explicit instantiations, while in the header-only case, you will not.

Reply:
Yes, but I think it's really only 2 or 3 types.

Reply:
But what if a user has his/her own `VectorType` or `MatrixType`? It won't be able to use the solver at all unless it copy-pastes the `.cc` code and recompile for the new classes.